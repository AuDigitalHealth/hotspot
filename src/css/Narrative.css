/*
 * Mandatory narrative styling from the FHIR specification.
 * See: http://hl7.org/fhir/STU3/narrative.html#css
 */
.narrative .bold {
  font-weight: bold;
}
.narrative .italics {
  font-style: italic;
}
.narrative .underline {
  text-decoration: underline;
}
.narrative .strikethrough {
  text-decoration: line-through;
}
.narrative .left {
  text-align: left;
}
.narrative .right {
  text-align: right;
}
.narrative .center {
  text-align: center;
}
.narrative .justify {
  text-align: justify;
}
.narrative .border-left {
  border-left: 1px solid grey;
}
.narrative .border-right {
  border-right: 1px solid grey;
}
.narrative .border-top {
  border-top: 1px solid grey;
}
.narrative .border-bottom {
  border-bottom: 1px solid grey;
}
.narrative .arabic {
  list-style-type: decimal;
}
.narrative .little-roman {
  list-style-type: lower-roman;
}
.narrative .big-roman {
  list-style-type: upper-roman;
}
.narrative .little-alpha {
  list-style-type: lower-alpha;
}
.narrative .big-alpha {
  list-style-type: upper-alpha;
}
.narrative .disc {
  list-style-type: disc;
}
.narrative .circle {
  list-style-type: circle;
}
.narrative .square {
  list-style-type: square;
}
.narrative .unlist {
  list-style-type: none;
}
