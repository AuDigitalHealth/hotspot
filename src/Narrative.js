import React, { Component } from 'react'
import PropTypes from 'prop-types'
import sanitize from 'sanitize-html'
import http from 'axios'
import omitBy from 'lodash.omitby'

import { translateHref } from './fhir/restApi.js'

import './css/Narrative.css'

// Renders the XHTML narrative content from a FHIR resource, optionally
// applying a custom stylesheet.
class Narrative extends Component {
  static propTypes = {
    content: PropTypes.string.isRequired,
    fhirServer: PropTypes.string.isRequired,
    stylesPath: PropTypes.string,
    onError: PropTypes.func,
  }

  constructor(props) {
    super(props)
    const state = {}
    if (props.content) {
      try {
        state.sanitizedContent = this.sanitizeContent(props.content)
      } catch (error) {
        if (props.onError) {
          props.onError(error)
        }
      }
    }
    this.state = state
  }

  async getStyles() {
    try {
      const response = await http.get(this.props.stylesPath)
      const contentType = response.headers['content-type']
      // Check that the Content-Type of the stylesheet response is `text/css`.
      if (!contentType.match(/text\/css/)) {
        throw Error(
          `Narrative stylesheet had unexpected format: "${contentType}"`,
        )
      }
      const styles = response.data
      this.setState(() => ({ styles }))
    } catch (error) {
      throw Error(`Narrative stylesheet not found: "${error.toString()}"`)
    }
  }

  sanitizeContent(content) {
    const { fhirServer } = this.props
    const allowedTags = require('./config/allowedTags.json')
    const sanitizedContent = sanitize(content, {
      // Allow only our set of whitelisted tags.
      allowedTags,
      // Allow all attributes (except for event-related, see later filter).
      allowedAttributes: false,
      // Allow data URLs.
      allowedSchemes: ['http', 'https', 'ftp', 'mailto', 'data'],
      transformTags: {
        // Filter out event-related attributes.
        '*': (tagName, attribs) => ({
          tagName,
          attribs: omitBy(attribs, (_, key) => key.match(/^on.+/)),
        }),
        // Translate any relative hrefs within anchors.
        a: (tagName, attribs) => ({
          tagName,
          attribs: {
            ...attribs,
            href: translateHref(attribs.href, fhirServer),
          },
        }),
      },
    })
    return sanitizedContent
  }

  componentDidMount() {
    if (this.props.stylesPath) {
      this.getStyles().catch(error => {
        if (this.props.onError) {
          this.props.onError(error)
        }
      })
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.content) {
      try {
        this.setState(() => ({
          sanitizedContent: this.sanitizeContent(nextProps.content),
        }))
      } catch (error) {
        nextProps.onError(error)
      }
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      this.props.content !== nextProps.content ||
      this.state.sanitizedContent !== nextState.sanitizedContent ||
      this.state.styles !== nextState.styles
    )
  }

  render() {
    const { sanitizedContent, styles } = this.state

    // Narrative styles are added to a "scoped" style tag inside the narrative
    // div. This means that these styles are only applied to content within
    // this element.
    const styleTag = styles ? <style scoped>{styles}</style> : null

    const narrative = sanitizedContent ? (
      <div
        className="narrative-content"
        dangerouslySetInnerHTML={{ __html: sanitizedContent }}
      />
    ) : (
      <p>This FHIR resource does not have narrative content.</p>
    )

    return (
      <div className="narrative">
        {styleTag}
        {narrative}
      </div>
    )
  }
}

export default Narrative
