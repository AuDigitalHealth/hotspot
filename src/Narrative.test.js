import React from 'react'
import { shallow } from 'enzyme'

import Narrative from './Narrative.js'

describe('Narrative', () => {
  it('should render correctly', () => {
    const props = {
      fhirServer: 'http://hl7.org/STU3',
      content: '<a href="http://somedomain.com/some/thing">Link</a>',
      stylesPath: '/some-styles.css',
    }
    const wrapper = shallow(<Narrative {...props} />)
    wrapper.setState({ styles: 'strong { font-weight: 500; }' })
    expect(wrapper).toMatchSnapshot()
  })

  describe('sanitization', () => {
    it('should filter out a tag not in the whitelist', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<div><audio src="music.mp3"/></div>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      const contentDiv = wrapper.find('.narrative-content').first()
      const content = contentDiv.prop('dangerouslySetInnerHTML')['__html']
      expect(content).toBe('<div></div>')
    })

    it('should allow data URLs', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<div><img src="data:image/gif;base64,c29tZWNyYXAK" /></div>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      const contentDiv = wrapper.find('.narrative-content').first()
      const content = contentDiv.prop('dangerouslySetInnerHTML')['__html']
      expect(content).toBe(props.content)
    })

    it('should filter out an onClick attribute', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content:
          '<div><a href="http://somedomain.com/some/thing" onclick="javascript:alert(\'Boo!\')">Link</a></div>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      const contentDiv = wrapper.find('.narrative-content').first()
      const content = contentDiv.prop('dangerouslySetInnerHTML')['__html']
      expect(content).toBe(
        '<div><a href="http://somedomain.com/some/thing">Link</a></div>'
      )
    })
  })

  describe('href translation', () => {
    const getLinkHref = wrapper => {
      const contentDiv = wrapper.find('.narrative-content').first()
      const contentDoc = new DOMParser().parseFromString(
        contentDiv.prop('dangerouslySetInnerHTML')['__html'],
        'application/xml'
      )
      const link = contentDoc.querySelector('a')
      return link.href
    }

    it('should not translate an absolute href', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="http://somedomain.com/some/thing">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe('http://somedomain.com/some/thing')
    })

    it('should not translate a protocol-relative href', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="//somedomain.com/some/thing">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe('//somedomain.com/some/thing')
    })

    it('should translate a root-relative href', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="/some/thing">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe('http://hl7.org/some/thing')
    })

    it('should translate a filename href', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="something.html">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe(
        'http://hl7.org/fhir/STU3/something.html'
      )
    })

    it('should translate a relative href (.)', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="./something">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe('http://hl7.org/fhir/STU3/something')
    })

    it('should translate a relative href (..)', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="../something">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe('http://hl7.org/fhir/something')
    })

    it('should not translate a fragment', () => {
      const props = {
        fhirServer: 'http://hl7.org/fhir/STU3',
        content: '<a href="#something">Link</a>',
      }
      const wrapper = shallow(<Narrative {...props} />)
      expect(getLinkHref(wrapper)).toBe('about:blank#something')
    })
  })
})
