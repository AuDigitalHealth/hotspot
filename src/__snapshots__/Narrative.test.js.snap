// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Narrative should render correctly 1`] = `
<div
  className="narrative"
>
  <style
    scoped={true}
  >
    strong { font-weight: 500; }
  </style>
  <div
    className="narrative-content"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<a href=\\"http://somedomain.com/some/thing\\">Link</a>",
      }
    }
  />
</div>
`;
